sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/DoNotOverwriteDictionaryValues.cs

```diff
@@ -0,0 +1,103 @@
+﻿using System;
+using System.Collections.Generic;
+
+namespace Tests.Diagnostics
+{
+    class TestCases
+    {
+        void Indexers(Dictionary<int, int> items)
+        {
+            int index;
+
+            items[index] = 1; // Secondary
+            items[index] = 2; // Noncompliant {{Verify this is the key that was intended; a value has already been set for it.}}
+//          ^^^^^^^^^^^^^^^^
+
+            items[0] = 1; // Secondary
+            items[1] = 2;
+            items[0] = 3; // Noncompliant
+
+            items[index++] = 1;
+            items[index++] = 2; // Compliant, non-const or variable indexes are ignored
+
+            items[2] = 1;
+            if (true)
+            {
+                items[2] = 2; // Compliant, it is conditional
+            }
+
+            items[3] = 1; // Secondary
+            items[3] = items[5]; // Noncompliant, not the same item is read
+        }
+
+        void Methods(Dictionary<int, int> items)
+        {
+            items.Add(0, 1); // Secondary
+            items.Add(0, 2); // Noncompliant
+
+            items[1] = 1; // Secondary
+            items.Add(1, 2); // Noncompliant
+
+            items.Add(2, 2); // Secondary
+            items[2] = 1; // Noncompliant
+        }
+
+        void Exceptions(Dictionary<int, int> items)
+        {
+            int index;
+
+            items[index] = 1;
+            index++; // any statement that does not contain element access or invocation breaks the search
+                     // for elements because currently looked up key is not const and could be modifying it
+            items[index] = 2; // Compliant, not subsequent lines
+
+            items[1] = 1; // Secondary
+            index++; // this does not break the search for elements because currently looked up key is const
+            items[1] = 2; // Noncompliant
+
+            items[2] = 1;
+            // existing item is used
+            items.Add(2, items[2]);
+
+            items[3] = 1;
+            // existing item is read
+            items[3] = items[3];
```

Comment:
Why is this line compliant? Is it a false-negative?

Reply:
It does not change the collection, perhaps we should raise (unless there is another rule that would raise in such cases) but we don't because we detect that the `items[3]` is read on the right hand side of the assignment.